src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs

```diff
@@ -107,27 +58,19 @@ public FileStream(IntPtr handle, FileAccess access, bool ownsHandle, int bufferS
                 throw;
             }
 
-            // Note: Cleaner to set the following fields in ValidateAndInitFromHandle,
-            // but we can't as they're readonly.
-            _access = access;
-            _useAsyncIO = isAsync;
-
-            // As the handle was passed in, we must set the handle field at the very end to
-            // avoid the finalizer closing the handle when we throw errors.
-            _fileHandle = safeHandle;
-        }
-
-        public FileStream(SafeFileHandle handle, FileAccess access)
-            : this(handle, access, DefaultBufferSize)
-        {
-        }
-
-        public FileStream(SafeFileHandle handle, FileAccess access, int bufferSize)
-            : this(handle, access, bufferSize, GetDefaultIsAsync(handle))
-        {
+            // it might seem to have no sense now, but we plan to introduce dedicated strategies for sync and async implementations
+            switch (isAsync)
+            {
+                case true:
+                    _impl = new FileStreamImpl(safeHandle, true, access, bufferSize, true);
+                    return;
+                case false:
+                    _impl = new FileStreamImpl(safeHandle, true, access, bufferSize, false);
```

Comment:
I'm surprised not to see `this` being passed in.  I would expect you'd need the parent FileStream reference in order to ensure that any calls methods make to other _virtual_ methods on the instance appropriately go through the parent `FileStream` in order to pick up any overrides of a derived class.  (If this shows up as a perf impact, in theory it's one of the things that can be specialized by one of the impls, with a GetType() == typeof(FileStream) check in the constructor to determine whether the impl needs to factor in a derived type's potential overrides.)

Reply:
this is an excellent point. I've realized that a few moments ago when I discovered that some `FileStream`-based tests from outside `System.IO.Tests` have started failing. I will fix this asap